controlplane/kubeadm/internal/proxy/dial.go

```diff
@@ -0,0 +1,160 @@
+	upgrader       spdy.Upgrader
+	timeout        time.Duration
+	requestIDLock  sync.Mutex
+	requestID      int
+}
+
+// NewDialer creates a new dialer for a given API server scope
+func NewDialer(p Proxy, options ...func(*Dialer) error) (*Dialer, error) {
+	if p.Port == 0 {
+		return nil, errors.New("port required")
+	}
+
+	dialer := &Dialer{
+		proxy: p,
+	}
+
+	for _, option := range options {
+		err := option(dialer)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if dialer.timeout == 0 {
+		dialer.timeout = defaultTimeout
+	}
+	p.KubeConfig.Timeout = dialer.timeout
+	clientset, err := kubernetes.NewForConfig(p.KubeConfig)
+	if err != nil {
+		return nil, err
+	}
+	proxyTransport, upgrader, err := spdy.RoundTripperFor(p.KubeConfig)
+	if err != nil {
+		return nil, err
+	}
+	dialer.proxyTransport = proxyTransport
+	dialer.upgrader = upgrader
+	dialer.clientset = clientset
+	return dialer, nil
+}
+
+// DialContextWithAddr is a GO grpc compliant dialer construct
+func (d *Dialer) DialContextWithAddr(ctx context.Context, addr string) (net.Conn, error) {
+	return d.DialContext(ctx, scheme, addr)
+}
+
+// nextRequestID generates the next request ID for calling to the API server
+func (d *Dialer) nextRequestID() int {
+	d.requestIDLock.Lock()
+	defer d.requestIDLock.Unlock()
+	id := d.requestID
+	d.requestID++
+	return id
+}
+
+// DialContext creates proxied port-forwarded connections.
+// ctx is currently unused, but fulfils the type signature used by GRPC.
+func (d *Dialer) DialContext(_ context.Context, network string, addr string) (net.Conn, error) {
+	req := d.clientset.CoreV1().RESTClient().
+		Post().
+		Resource(d.proxy.Kind).
+		Namespace(d.proxy.Namespace).
+		Name(d.proxy.ResourceName).
+		SubResource("portforward")
+
+	dialer := spdy.NewDialer(d.upgrader, &http.Client{Transport: d.proxyTransport}, "POST", req.URL())
+
+	p, _, err := dialer.Dial(portforward.PortForwardProtocolV1Name)
+	if err != nil {
+		return nil, errors.Wrap(err, "error upgrading connection")
+	}
+	headers := http.Header{}
+	headers.Set(corev1.StreamType, corev1.StreamTypeError)
+	headers.Set(corev1.PortHeader, fmt.Sprintf("%d", d.proxy.Port))
+	headers.Set(corev1.PortForwardRequestIDHeader, strconv.Itoa(d.nextRequestID()))
+	errorStream, err := p.CreateStream(headers)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := errorStream.Close(); err != nil {
+		return nil, err
+	}
+
+	errorChan := make(chan error, 2)
+	go func() {
+		message, err := ioutil.ReadAll(errorStream)
+		switch {
+		case err != nil:
+			errorChan <- errors.Wrap(err, "error reading from error stream")
+		case len(message) > 0:
+			errorChan <- errors.Wrapf(err, "an error occurred forwarding: %v", string(message))
+		}
+		close(errorChan)
+	}()
+
+	headers.Set(corev1.StreamType, corev1.StreamTypeData)
+	dataStream, err := p.CreateStream(headers)
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating forwarding stream")
+	}
+
+	c := NewConn(dataStream, errorChan)
```

Comment:
How do you see errorChan being used by a consumer? This func returns a `net.Conn`, so someone would have to type assert it to our `proxy.Conn` and then read from it somehow/somewhere...

Reply:
Also came to mind 20 mins after posting. Not sure how to track it.

I've tested that if it does error, the connection terminates and EOFs rather than cause a panic, but not sure how to improve.

Reply:
Maybe we just get rid of trying to read from `errorStream`?

Reply:
Done